File: Features/NanoDo/Models/NanoDo.swift
//
//  NanoDo.swift
//  ToDo
//
//  Created by Moinuddin Ahmad on 2/9/26.
//

import Foundation
import SwiftData

@Model
final class NanoDo {
    #Index<NanoDo>([\.ownerUserID])

    var cloudID: UUID? = nil
    var ownerUserID: UUID? = nil
    var task: String = ""
    var createdAt: Date = Date()
    var updatedAt: Date? = nil
    var dueDate: Date? = nil
    var isDone: Bool = false

    var toDo: ToDo? = nil
    @Relationship(inverse: \Tag.nanoDos)
    var tag: Tag? = nil

    init(
        task: String,
        createdAt: Date = Date(),
        updatedAt: Date? = nil,
        dueDate: Date? = nil,
        isDone: Bool = false,
        toDo: ToDo? = nil,
        tag: Tag? = nil,
        cloudID: UUID? = nil,
        ownerUserID: UUID? = nil
    ) {
        self.cloudID = cloudID
        self.ownerUserID = ownerUserID
        self.task = task
        self.createdAt = createdAt
        self.updatedAt = updatedAt ?? createdAt
        self.dueDate = dueDate
        self.isDone = isDone
        self.toDo = toDo
        self.tag = tag
    }

    var syncUpdatedAt: Date {
        updatedAt ?? createdAt
    }

    func markUpdated(_ date: Date = .now) {
        updatedAt = date
    }
}

File: ToDo-Android/app/src/main/java/dev/iamshift/todo/android/core/model/NanoDo.kt
package dev.iamshift.todo.android.core.model

import java.time.Instant
import java.util.UUID

/** A child action belonging to one ToDo. The Supabase column is named `task`. */
data class NanoDo(
    val id: UUID = UUID.randomUUID(),
    val todoId: UUID,
    val ownerUserId: UUID? = null,
    val task: String,
    val isDone: Boolean = false,
    val tagId: UUID? = null,
    val dueAt: Instant? = null,
    val createdAt: Instant = Instant.now(),
    val updatedAt: Instant = createdAt
) {
    init {
        require(task.isNotBlank()) { "NanoDo task must not be blank" }
    }

    val syncUpdatedAt: Instant
        get() = updatedAt

    fun transitionToDone(done: Boolean, at: Instant = Instant.now()): NanoDo =
        copy(isDone = done, updatedAt = at)

    fun withTask(value: String, at: Instant = Instant.now()): NanoDo {
        val normalized = value.trim()
        require(normalized.isNotEmpty()) { "NanoDo task must not be blank" }
        return copy(task = normalized, updatedAt = at)
    }

    fun withDueAt(value: Instant?, at: Instant = Instant.now()): NanoDo =
        copy(dueAt = value, updatedAt = at)

    fun withTagId(value: UUID?, at: Instant = Instant.now()): NanoDo =
        copy(tagId = value, updatedAt = at)
}
